File: lab/test/art_test.clj
(ns art-test
  (:require [clojure.test :refer [deftest is testing]]
            [clojure.data.json :as json]
            [clojure.java.io :as io]
            [art :refer [mulberry32 generate-static-params]]))

(deftest test-mulberry32-vectors
  (testing "Clojure mulberry32 matches frozen test vectors for seeds 1, 42, 99"
    (let [data (json/read-str (slurp (io/resource "vectors.json")) :key-fn identity)
          vectors (get data "rng")]
      (doseq [[seed-str expected-vals] vectors]
        (let [seed (Long/parseLong seed-str)
              rng (mulberry32 seed)
              generated (repeatedly (count expected-vals) rng)]
          (doseq [[gen exp] (map vector generated expected-vals)]
            (is (< (Math/abs (- (double gen) (double exp))) 1e-9))))))))

(deftest test-static-params-vectors
  (testing "Clojure generate-static-params matches frozen vectors for seeds 1, 42, 99"
    (let [data (json/read-str (slurp (io/resource "vectors.json")) :key-fn identity)
          vectors (get data "staticParams")]
      (doseq [[seed-str expected] vectors]
        (let [seed (Long/parseLong seed-str)
              params (generate-static-params seed)
              exp-count (long (get expected "count"))
              exp-twist (double (get expected "twist"))
              exp-scales (vec (map double (get expected "scales")))
              exp-palette (long (get expected "paletteIdx"))
              exp-light (double (get expected "lightAngle"))]
          (is (= (:count params) exp-count))
          (is (< (Math/abs (- (:twist params) exp-twist)) 1e-9))
          (is (= (:paletteIdx params) exp-palette))
          (is (< (Math/abs (- (:lightAngle params) exp-light)) 1e-9))
          (is (= (count (:scales params)) (count exp-scales)))
          (doseq [[gen exp] (map vector (:scales params) exp-scales)]
            (is (< (Math/abs (- (double gen) (double exp))) 1e-9))))))))

File: web/test/art.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { mulberry32, generateRandomSeed } from '../src/rng.ts';
import { generateStaticParams, draw } from '../src/art.ts';

const vectorsPath = join(import.meta.dirname, 'vectors.json');
const vectors = JSON.parse(readFileSync(vectorsPath, 'utf8'));

describe('RNG mulberry32', () => {
  it('matches frozen test vectors for seeds 1, 42, 99', () => {
    for (const [seedStr, expectedValues] of Object.entries(vectors.rng)) {
      const seed = parseInt(seedStr, 10);
      const rng = mulberry32(seed);
      const generated = (expectedValues as number[]).map(() => rng());
      
      generated.forEach((val, idx) => {
        const expected = (expectedValues as number[])[idx];
        assert.ok(
          Math.abs(val - expected) < 1e-9,
          `Seed ${seed} index ${idx}: expected ${expected}, got ${val}`
        );
      });
    }
  });

  it('generates random seeds within [0, 99999]', () => {
    for (let i = 0; i < 50; i++) {
      const seed = generateRandomSeed();
      assert.strictEqual(Number.isInteger(seed), true);
      assert.strictEqual(seed >= 0, true);
      assert.strictEqual(seed <= 99999, true);
    }
  });
});

describe('Art Parameter Generator', () => {
  it('matches frozen staticParams vectors for seeds 1, 42, 99', () => {
    for (const [seedStr, expected] of Object.entries(vectors.staticParams)) {
      const seed = parseInt(seedStr, 10);
      const params = generateStaticParams(seed);
      const exp = expected as any;

      assert.strictEqual(params.count, exp.count);
      assert.ok(Math.abs(params.twist - exp.twist) < 1e-9);
      assert.strictEqual(params.paletteIdx, exp.paletteIdx);
      assert.ok(Math.abs(params.lightAngle - exp.lightAngle) < 1e-9);

      assert.strictEqual(params.scales.length, exp.scales.length);
      params.scales.forEach((scale, idx) => {
        assert.ok(Math.abs(scale - exp.scales[idx]) < 1e-9);
      });
    }
  });

  it('draw returns frame transform', () => {
    const params = generateStaticParams(42);
    const frame = draw(params, 1.0);
    assert.strictEqual(typeof frame.rotationY, 'number');
    assert.strictEqual(typeof frame.scaleFactor, 'number');
    assert.strictEqual(typeof frame.lightAngle, 'number');
  });
});
